### DISCUSSION TITLE
Filtering seemingly not able to properly connect logic with "&&" operator

### DISCUSSION THREAD

**Discussion by amandoee** [pocketbase/pocketbase#5223] [Q&A]
Hi, i've been trying to get my custom filtering to work with pocketbase, using Flutter.

I've followed the documentation, and have succesfully created a filters that work as intended.  I am creating a database with Clubs.

There are a few collections:
Club (Which is just recordid and logo of club)
ClubTags (Which is a collection of single-to-multiple relations from a club to multiple Tags)
Tags (Which is just id and tag String)

The problem arises with one of my filters. I wish to filter only RecordModels of Clubs where a given Club's tags are all present in the List of selected tag Strings.

No Problem right? 

I construct the filter as such:

if (interestTags.isNotEmpty) {
      String matchesInterest = '(';
      matchesInterest += interestTags.map((tag) => "ClubTags_via_club.clubTag.tag?='${tag}'").join(' && ');
      matchesInterest += ')';
      completeFilterList.add(matchesInterest);
    }

//Other filters [...]

return completeFilterList.join(' && ');

When i select a single interestTag, it works like a charm. When i select multiple however, i don't get any results. I am not getting an error, the query is simply not returning any results. Multiple entries in my collections should satisfy the criteria of being from both. 

I tried to add ":every", but it seems that the ":every" does not inherit members like ".tag", since i get an error: 
matchesInterest += interestTags.map((tag) => "ClubTags_via_club.clubTag:every.tag?='${tag}'").join(' && ');

Has anyone had a similar problem, and if so, is it possible to achieve what i am trying to do? I suspect i might have to change the structure of my database, but please let me know.

**Comment by ganigeorgiev** [comment]
Similar as in a SQL `WHERE` expression, a single row field value cannot be 2 things at the same time.
Also note that `:every` is not existing and valid modifier.

I'm having difficulties understanding your described collections structure to provide a more concrete example, but you can find a workaround for a similar problem in https://github.com/pocketbase/pocketbase/discussions/4319.

Alternatively, if you want to target "ClubTags" that has at least one matching tag from the list you can replace the `&&` with `||` for  a `IN` like query.

If the above doesn't help, please export your collections configuration as json (_Admin UI > Settings > Export collections_).

**Reply by amandoee** [reply]
Thanks for your quick answer!
I of course meant ":each" originally. :)

I don't believe the workaround will be able to work in my example, but here is my collection configuration attached.

[pb_schema.json](https://github.com/user-attachments/files/16249031/pb_schema.json)

**Reply by ganigeorgiev** [reply]
Based on the provided json and your description, the use case seems the same as in the linked discussion.

If you want to check for inclusion of all tags in a `ClubTags.clubTag` multiple `relation` field, then you could use the `~` operator to filter using the tag ids (if you don't know the tag ids you can perform an extra query before your main one as shown in the example in https://github.com/pocketbase/pocketbase/discussions/4319).

Or in other words a filter like this should work:

```dart
final tagIds = ["id1", "id2"];

final records = await pb.collection('Club').getFullList(
  // ClubTags_via_club.clubTag?~'"id1"' && ClubTags_via_club.clubTag?~'"id2"'
  filter: tagIds.map((id) => 'ClubTags_via_club.clubTag?~\'"$id"\'').join("&&");
);
```

The above should return all `Club` collection records that has at least one `ClubTags` relation that contains both "id1" and "id2" `Tags`.

_(the `?` prefix is used because back-relations resolve to a dynamic multiple `relation` field unless there is a `UNIQUE` constraint on the field; see https://pocketbase.io/docs/working-with-relations/#back-relation-caveats)_

**Reply by amandoee** [reply]
Thanks again. I tried the code, and it did unfortunately not work.
Upon attempt, it does not return any club, even if only a single tag is in the list.

**Reply by ganigeorgiev** [reply]
> Upon attempt, it does not return any club, even if only a single tag is in the list.

I'm confused and no longer sure what you are trying to do. The above query will return Club records **only** if both tag ids are present. If you want any/at-least-one type of comparision then using `ClubTags_via_club.clubTag.tag ?= 'something1' || ClubTags_via_club.clubTag.tag ?= 'something2'` should work.

In any case without an example how you are sending the request there is nothing much I can help further with and I will stop engaging to avoid further confusion.

**Reply by amandoee** [reply]
I wish to do exactly as you stated in your earlier reply: ("The above should return all Club collection records that has at least one ClubTags relation that contains both "id1" and "id2" Tags.").

I agree that i believe the code should work as you state, judging by the documentation. But the result is simply not what is to be expected.  Below is my logic:

`if (interestTags.isNotEmpty) {
      String matchesInterest = '(';

      matchesInterest += interestTags.map((id) => "ClubTags_via_club.clubTag.tag ?~'$id'").join(" && ");

      matchesInterest += ')';
      completeFilterList.add(matchesInterest);
    }`

When filtering with `interestTags.length==1`, the results show all Club collection models where the ClubTags relation contains that at least contain that id - this is as i would expect.

When filtering with `interestTags.length==2`, the results are empty. This is even though i have a ClubTags relation that contains both id's present in the `interestTags` List.

Does my problem make sense? Otherwise i would be happy to clarify!

**Reply by ganigeorgiev** [reply]
But you are not doing the same thing. `ClubTags_via_club.clubTag.tag` refers to the `tag` text field value (aka. name), not its id. Please copy the exact filter as shown in the example in https://github.com/pocketbase/pocketbase/discussions/5223#discussioncomment-10062537.

**Comment by Sigerolem** [comment]
I just had a, not similar, but maybe related problem.
When doing filters with '&&' pocketbase did not warn me, but, I was having errors.
I noticed that when I used '&&' instead of '||', it was not working correctly. It behave as '||'
I solved it by replacing '&&' on my query with '%26%26'
Maybe this will help you.
